### PR TITLE
[system-command-line] help implementation

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
https://github.com/dotnet/templating/issues/3810

### Solution
Implements help for new parser

TODO: 
- [x] revise command, option and argument descriptions
- [x] unit tests
- [x] clean todos in the code

Error handling for other cases will be done together with instantiation error handling in another PR.
Open question: 
- is it possible to fail help?  
- is it possible to use colors in help? - yes.

### Checks:
- [x] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Cli/Commands/InstallCommand.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using System.CommandLine;
 using System.CommandLine.Invocation;
 using System.CommandLine.Parsing;
+using System.Net.Mail;
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstallCommand.cs:9]: TODO: remove

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs [hunk 2]
```diff
@@ -120,45 +132,61 @@ protected async override Task<NewCommandStatus> ExecuteAsync(InstantiateCommandA
...
+                instantiateArgs,
+                environmentSettings,
+                templatePackageManager,
+                selectedTemplateGroups.Single(),
+                cancellationToken).ConfigureAwait(false);
         }
 
         protected override InstantiateCommandArgs ParseContext(ParseResult parseResult) => new(this, parseResult);
 
+#pragma warning disable SA1202 // Elements should be ordered by access
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs:162]: TODO: fix

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs [hunk 3]
```diff
@@ -181,6 +209,11 @@ protected async override Task<NewCommandStatus> ExecuteAsync(InstantiateCommandA
             return NewCommandStatus.NotFound;
         }
 
+        //TODO
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.cs:212]: TODO: fix

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+using Microsoft.TemplateEngine.Cli.Extensions;
+using Microsoft.TemplateEngine.Cli.TabularOutput;
+using Microsoft.TemplateEngine.Edge.Settings;
+using Microsoft.TemplateEngine.Utils;
+
+namespace Microsoft.TemplateEngine.Cli.Commands
+{
+    internal class InstantiateHelpBuilder : HelpBuilder
+    {
+        //TODO: remove if https://github.com/dotnet/command-line-api/pull/1484/files is merged
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs:21]: https://github.com/dotnet/command-line-api/pull/1484

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+                base.Write(command, writer, parseResult);
+                return;
+            }
+
+            IEngineEnvironmentSettings environmentSettings = instantiateCommand.GetEnvironmentSettingsFromArgs(instantiateCommandArgs);
+
+            using TemplatePackageManager templatePackageManager = new TemplatePackageManager(environmentSettings);
+            HostSpecificDataLoader hostSpecificDataLoader = new HostSpecificDataLoader(environmentSettings);
+
+            //TODO: consider use cache only for help
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs:45]: will be done in separate issue

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+                            parametersToShow[param.Name] = new BaseTemplateParameter(param, aliases);
+                        }
+
+                    }
+                }
+            }
+            return parametersToShow.Values;
+        }
+
+        //TODO: remove if https://github.com/dotnet/command-line-api/pull/1484/files is merged
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs:260]: https://github.com/dotnet/command-line-api/pull/1484

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+                                displayValue.Append("- " + choiceInfo.Value);
+                            }
+
+                            displayValue.AppendLine();
+                        }
+                    }
+                    else
+                    {
+                        displayValue.Append(param.DataType ?? "string");
+                        //TODO: localize
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateHelpBuilder.cs:467]: fix

File: src/Microsoft.TemplateEngine.Cli/Commands/TemplateCommand.cs [hunk 8]
```diff
@@ -166,6 +176,9 @@ public async Task<int> InvokeAsync(InvocationContext context)
             }
         }
 
+        //TODO: implement this
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/TemplateCommand.cs:179]: will be done in separate PR

File: src/dotnet-new3/ParserFactory.cs [hunk 9]
```diff
@@ -19,14 +21,16 @@ internal static Parser CreateParser(Command command, bool disableHelp = false)
...
-            //    var hb = new HelpBuilder(LocalizationResources.Instance);
-            //    hb.Customize(command, (result) => )
-            //    return hb;
-            // })
+            .UseParseErrorReporting() //TODO: discuss with SDK if it is possible to use it.
+            .UseHelpBuilder(context =>
+            {
+                if (context.ParseResult.CommandResult.Command is ICustomHelpBuilder custom)
+                {
+                    //TODO: create IEnvironment early and use IEnvironment property for that
```
Comment (vlada-shubina) [inline src/dotnet-new3/ParserFactory.cs:29]: will be done in separate PR

File: test/dotnet-new3.UnitTests/DotnetNewHelp.Approval.cs [hunk 10]
```diff
@@ -79,7 +166,7 @@ public void CanShowHelpForTemplate_Classlib(string command)
         }
 
 #pragma warning disable xUnit1004 // Test methods should not be skipped
-        [Fact(Skip = "Help is not implemented yet")]
+        [Fact(Skip = "TODO: does not fail now, check if can fail")]
```
Comment (vlada-shubina) [inline test/dotnet-new3.UnitTests/DotnetNewHelp.Approval.cs:169]: To discuss, whether help can return 1 exit code at all.

**Comment by vlada-shubina** [comment]
@baronfel though this is still a draft, could you please start reviewing `.approved.txt`s in this PR to see if there are more changes needed to layouts/wording? For now, I would appreciate feedback on layout first, as this may need code change - wording can be changed any time in resx.

We are still discussing the implementation details, however the layouts should not change. 
Web project templates are included just for reference - they have more parameters and 3rd party notice - please skip wording in template options there.

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp_List.approved.txt [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+Arguments:
+  <name>  If specified, only the templates matching the name will be shown.
+
+Options:
+  --author <author>                      Filters the templates based on the template author.
+  -lang, --language <language>           Filters templates based on language.
+  --type <type>                          Filters templates based on available types. Predefined values are "project" and "item".
+  --tag <tag>                            Filters the templates based on the tag.
+  --columns-all                          Display all columns in the output.
+  --columns <author|language|tags|type>  Specifies the columns to display in the output.
```
Comment (baronfel) [inline test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp_List.approved.txt:16]: random thought not directly part of this rewrite - why `--columns-all` instead of `--columns all`. making `all` just another display option.

File: test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt [hunk 12]
```diff
@@ -1,8 +1,22 @@
 ﻿ASP.NET Core Web App (Model-View-Controller) (C#)
 Author: Microsoft
-Description: A project template for creating an ASP.NET Core application with example ASP.NET Core MVC Views and Controllers. This template can also be used for RESTful HTTP services.
+Description: A project template for creating an ASP.NET Core application with example ASP.NET Core MVC Views and Controllers. This template can also be used
```
Comment (baronfel) [inline test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt:3]: there's an odd line break here, is this just because of the line width difference between the old and the new formats?  When looking at these tests, is there a given line width to keep the output consistent?

File: test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt [hunk 13]
```diff
@@ -1,8 +1,22 @@
 ﻿ASP.NET Core Web App (Model-View-Controller) (C#)
 Author: Microsoft
-Description: A project template for creating an ASP.NET Core application with example ASP.NET Core MVC Views and Controllers. This template can also be used for RESTful HTTP services.
+Description: A project template for creating an ASP.NET Core application with example ASP.NET Core MVC Views and Controllers. This template can also be used 
+for RESTful HTTP services.
 This template contains technologies from parties other than Microsoft, see https://aka.ms/aspnetcore/5.0-third-party-notices for details.
-Options:                                                                                                                                                                                   
+
+Usage:
+  new3 mvc [options] [template options]
```
Comment (baronfel) [inline test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt:8]: I really like the way the template-specific help is laid out here.  I like the named sections in the example usage, and the symmetry with the section names in the detailed output.

The formatting of the options themselves is also good. Variants, defaults, and parameter names are all clearly distinguishable.

File: test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt [hunk 14]
```diff
@@ -13,43 +27,43 @@ Options:
                                     Default: None                                                                                                                                          
 
   --aad-b2c-instance                The Azure Active Directory B2C instance to connect to (use with IndividualB2C auth).                                                                   
-                                    string - Optional                                                                                                                                      
+                                    text - Optional
```
Comment (baronfel) [inline test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt:30]: do you know why there's a trailing space on these examples?

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt [hunk 15]
```diff
@@ -1,25 +1,16 @@
...
-  --search <PARTIAL_NAME>        Searches for the templates on NuGet.org.
-  --author <AUTHOR>              Filters the templates based on the template author. Applicable only with --search or --list | -l option.
-  --package <PACKAGE>            Filters the templates based on NuGet package ID. Applies to --search.
-  --columns <COLUMNS_LIST>       Comma separated list of columns to display in --list and --search output.
-                                 The supported columns are: language, tags, author, type.
-  --columns-all                  Display all columns in --list and --search output.
-  --tag <TAG>                    Filters the templates based on the tag. Applies to --search and --list.
-  --no-update-check              Disables checking for the template package updates when instantiating a template.
\ No newline at end of file
+Usage:
```
Comment (baronfel) [inline test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt:4]: This help output is laid out a little oddly to my eye.

The usage suggests that a command could come after the template short name and template args, `dotnet new3 webapi --langVersion latest update`, but that is not valid, is it?

In cases like these, where a root command has multiple branching usages, I've seen some CLI tools offer examples of the two modes separately:

```
Usage: 
new3 [<template-short-name> [<template-args>...]]
new3 [command]
```

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp_Install.approved.txt [hunk 16]
```diff
@@ -0,0 +1,15 @@
+﻿Description:
+  Installs a template package.
+
+Usage:
+  new3 install <package>... [options]
```
Comment (baronfel) [inline test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp_Install.approved.txt:5]: does `...` in this usage imply repetition?

**Review by baronfel — COMMENTED** [review]
I really like the new formatting overall, it's very easy to understand. I left a few specific notes inline.

File: test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt (hunk 12 above)
Comment (vlada-shubina) [reply test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt:3]: I'll check it. `HelpBuilder` has `MaxWidth` however it should be used when the output is redirected. These also may change with alternative help customization approach.

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt (hunk 15 above)
Comment (vlada-shubina) [reply test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt:4]: yes, both template and subcommand is not valid. I will do custom usage here as above.

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp_Install.approved.txt (hunk 16 above)
Comment (vlada-shubina) [reply test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp_Install.approved.txt:5]: yes, it means 1 or more argument.

File: test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt (hunk 12 above)
Comment (vlada-shubina) [reply test/dotnet-new3.UnitTests/Approvals/AllWebProjectsWork.CanShowHelp_Mvc.approved.txt]: not applicable anymore

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt (hunk 15 above)
Comment (vlada-shubina) [reply test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt:4]: fixed

Comment (vlada-shubina) [reply test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt:4]: @baronfel do you prefer to keep "options" or remove them below?

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt [hunk 17]
```diff
@@ -1,22 +1,35 @@
...
+  -n, --name <name>       The name for the output being created. If no name is specified, the name of the output directory is used.
+  -o, --output <output>   Location to place the generated output.
+  --dry-run               Displays a summary of what would happen if the given command line were run if it would result in a template creation.
+  --force                 Forces content to be generated even if it would change existing files.
+  --no-update-check       Disables checking for the template package updates when instantiating a template.
+  -lang, --language <C#>  Specifies the template language to instantiate.
+  --type <project>        Specifies the template type to instantiate.
 
+Template options:
+  -f, --framework <net5.0|net7.0|netcoreapp3.1|netstandard2.0|netstandard2.1>  The target framework for the project.
```
Comment (vlada-shubina) [inline test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt:18]: @baronfel how do you feel about this notation? or should we rather not list possible options for choices?

now it makes 1st column really large

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs [hunk 18]
```diff
@@ -0,0 +1,473 @@
...
+                    break;
+                }
+            }
+
+            //build new context with helper command
+            var newContext = new HelpContext(context.HelpBuilder, optionsSetCommand, context.Output, null, context.MaxWidth);
+
+            //show option section
+            HelpBuilder.OptionsSection()(newContext);
+            context.Output.WriteLine();
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs:470]: @jonsequitur is there any better approach to print certain list of options?

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt (hunk 17 above)
Comment (vlada-shubina) [reply test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt:18]: also, for required arguments: the syntax of first column is `--my-choice <val1|val2|val3|val4> required`.

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt (hunk 15 above)
Comment (baronfel) [reply test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelp.approved.txt:4]: Options works great here 👍

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt (hunk 17 above)
Comment (baronfel) [reply test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt:18]: I've been doing some digging around that, trying out sample code for popular golang/nodejs/python CLI libraries to compare. So far they either:

* put the choices in a longer documentation section (like you've shown with the TFM documentation section),
* after a certain line width, put the documentation on a new line and ignore the long line for purposes of indentation on other lines, or 
* sidestep the issue entirely by forcing users to write all usage/help documentation manually

of those options, in an ideal world I'd like to have a behavior like:

* start by doing the current inline behavior
* if the line gets too long (we can debate what too long means, but let's say ~50 of terminal width), put the documentation on a new line compared to the flag usage example

or a behavior like:
* if there are more than N choices, do not show the choices in line with the argument. instead, just rely on the full listing of the options in the help documentation.

I'm not sure how feasible something like that is in the current S.CL API, partially because at least part of that seems like a terminal concern, not a CLI app concern necessarily. This is where something like GNU readline would be useful.

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.HelpUtils.cs [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+
+#nullable enable
+
+using System.CommandLine;
+using System.CommandLine.Help;
+using System.Text;
+
+namespace Microsoft.TemplateEngine.Cli.Commands
+{
+    internal partial class InstantiateCommand
```
[the remark this answers is not in the document]
Comment (jonsequitur) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.HelpUtils.cs:12]: Because we got rid of the `IHelpBuilder` interface, I'm inclined to make `GetTwoColumnRow` public in-place. It has dependencies on other `HelpBuilder` state (such as per-symbol customizations) so this would be the simplest thing.

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs [hunk 20]
```diff
@@ -0,0 +1,473 @@
...
+                    templatePackageManager,
+                    templateGroup);
+
+            if (!VerifyMatchingTemplates(
+                environmentSettings,
+                context.Output,
+                matchingTemplates,
+                out IEnumerable<TemplateCommand>? templatesToShow))
+            {
+                //error
```
[the remark this answers is not in the document]
Comment (jonsequitur) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs:76]: Under what circumstances would help fail?

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs [hunk 21]
```diff
@@ -0,0 +1,473 @@
...
+            CustomUsageSection(context);
+            HelpBuilder.CommandArgumentsSection()(context);
+            context.Output.WriteLine();
+            HelpBuilder.OptionsSection()(context);
+            context.Output.WriteLine();
+            HelpBuilder.SubcommandsSection()(context);
+            context.Output.WriteLine();
+            HelpBuilder.AdditionalArgumentsSection()(context);
+            context.Output.WriteLine();
+        }
```
[the remark this answers is not in the document]
Comment (jonsequitur) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs:271]: Is the pattern of setting `HelpBuilder.Layout` via `UseHelp` and returning an `IEnumerable<HelpDelegate>` not working for you?

File: src/dotnet-new3/ParserFactory.cs [hunk 22]
```diff
@@ -20,22 +22,34 @@ internal static Parser CreateParser(Command command, bool disableHelp = false)
...
+                {
+                    foreach (var section in HelpBuilder.DefaultLayout())
+                    {
+                        section.Invoke(context);
+                        context.Output.WriteLine();
+                    }
+                }
+            };
+        }
+
```
[the remark this answers is not in the document]
Comment (jonsequitur) [reply src/dotnet-new3/ParserFactory.cs:52]: Maybe `UseHelp` should accept a `Func<HelpContext, IEnumerable<HelpDelegate>>` rather than  `IEnumerable<HelpDelegate>` ? This way you could hoist the decision up and do something like this: 

```cs
if (context.ParseResult.CommandResult.Command is ICustomHelp custom)
{
    return CustomLayout();
}
else
{
    return context.HelpBuilder.DefaultLayout();
}
```

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs:76]: This is related to our "reparsing" thing. 
Examples may be:
`dotnet new bla -h` - `bla` template doesn't exist, we cannot show the help
`dotnet new my-multi-language-template` - we can only show help for single language, and user didn't specify one. 

There are more complicated cases. 

In fact this is minor imo. Previously in these cases `dotnet new` returned error exit code, but now it will not, but still show this message. Also we cannot use colors in help output which is a pity, but also minor.

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs:271]: not in this case. 
We have custom help for instantiate command already:
- if argument (template name was not given) we need to use standard output with custom usage (this method).
- if argument is given. the other flow is run.

[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.Help.cs:271]: If [this ](https://github.com/dotnet/templating/pull/4143#discussion_r766182868) is done, it is improves this problem too.

File: src/Microsoft.TemplateEngine.Cli/ChoiceTemplateParameter.cs [hunk 23]
```diff
@@ -25,20 +27,39 @@ internal ChoiceTemplateParameter(ITemplateParameter parameter, HostSpecificTempl
...
+            _choices = choiceTemplateParameter.Choices.ToDictionary(kvp => kvp.Key, kvp => kvp.Value, StringComparer.OrdinalIgnoreCase);
         }
 
         internal override ParameterType Type => ParameterType.Choice;
 
-        internal virtual IReadOnlyDictionary<string, ParameterChoice> Choices => RawChoices;
+        internal virtual IReadOnlyDictionary<string, ParameterChoice> Choices => _choices;
 
-        protected IReadOnlyDictionary<string, ParameterChoice> RawChoices { get; }
+        //TODO: needs https://github.com/dotnet/command-line-api/pull/1527
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Cli/ChoiceTemplateParameter.cs:49]: needs https://github.com/dotnet/command-line-api/pull/1527

File: test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt [hunk 24]
```diff
@@ -1,22 +1,36 @@
...
+                                                                               Type: choice
+                                                                                 net7.0          Target net7.0
```
Comment (vlada-shubina) [inline test/dotnet-new3.UnitTests/Approvals/DotnetNewHelp.CanShowHelpForTemplate_Classlib.approved.txt:20]: @baronfel another question here: now choices in description seems not ordered.
In fact they are ordered by template precedence and then by order given in templates. Do we want to align ordering in first and second columns? This way the first option might come from less precedence template.

File: src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.HelpUtils.cs [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        //TODO: this code is from System.CommandLine. Ideally if GetUsageParts (get first column by symbol) is exposed, we don't need it.
```
Comment (bekir-ozturk) [inline src/Microsoft.TemplateEngine.Cli/Commands/InstantiateCommand.HelpUtils.cs:14]:
```suggestion
        // TODO: this code is from System.CommandLine. Ideally if GetUsageParts (get first column by symbol) is exposed, we don't need it.
        // https://github.com/dotnet/command-line-api/blob/98dbbecec3d10698950e091a65a68243e36530db/src/System.CommandLine/Help/HelpBuilder.cs#L198
```

File: src/Microsoft.TemplateEngine.Cli/Commands/AliasAddCommand.cs [hunk 26]
```diff
@@ -29,7 +29,10 @@ internal LegacyAliasAddCommand(ITemplateEngineHost host, ITelemetryLogger logger
     internal class BaseAliasAddCommand : BaseCommand<AliasAddCommandArgs>
     {
         internal BaseAliasAddCommand(ITemplateEngineHost host, ITelemetryLogger logger, NewCommandCallbacks callbacks, string commandName)
-            : base(host, logger, callbacks, commandName) { }
+            : base(host, logger, callbacks, commandName)
+        {
+            Description = SymbolStrings.Command_AliasAdd_Description;
+        }
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Cli/Commands/AliasAddCommand.cs:35]: in theory, description is optional, but if not given, it will be missing in help. 
I will make it mandatory for our commands not to miss it.
